extension/lib/doc.js

```diff
@@ -181,45 +190,54 @@ class Doc {
                     termFreq[word]++;
                 } else {
                     termFreq[word] = 1;
-                }
+                };
             };
             if ( wordList) { total_words += wordList.length; };
         }
 
-        // Update term document frequency object: 
-        settings.getTermDocumentFreq(function(settingsTermDocumentFreq) {
-            termDocumentFreq = settingsTermDocumentFreq;
-            // This console.log is for monitoring the total word count as I go, to see how it progresses
-            debug("Number of words in document frequency dictionary: "+Object.keys(termDocumentFreq).length);
+        // Get/update document frequency dict, set keywords
+        settings.getDocumentFreq(function(settingsDocumentFreq) {
+            documentFreq = settingsDocumentFreq;
+            // This debug is for monitoring the total word count as I go, to see how it progresses
+            debug("Number of words in document frequency dictionary: "+Object.keys(documentFreq).length);
             settings.getVisitedUrls(function(settingsVisitedUrls) {
                 let visitedUrls = settingsVisitedUrls;
-                if (visitedUrls[window.location]) { // Check if been to site before
-                    // Do nothing
-                } else {
-                    self.updateTermDocumentFreq(termFreq, termDocumentFreq, settings.setTermDocumentFreq);
+                let num_documents = Object.keys(visitedUrls).length;
+                // Determine keywords BEFORE update document freq. -> save time, shouldn't affect result much
+                self.setKeyWords(termFreq, documentFreq, num_documents);
+                // Check if haven't been to site before
+                if (!visitedUrls[window.location]) {
+                    self.updateDocumentFreq(termFreq, documentFreq, function() {
+                        settings.setDocumentFreq
+                    });
                     visitedUrls[window.location] = 1;
                     settings.setVisitedUrls(visitedUrls);
                 }
-                let num_docs = Object.keys(visitedUrls).length;
-                self.num_docs = num_docs;
             })
         });
 
         this.termFreq = termFreq; // Set class attribute "termFreq"
-        return total_words // Set total words for use elsewhere (like Display)
+        this.total_words = total_words; // Set total words for use elsewhere (like Display)
     }
 
-    updateTermDocumentFreq(termFreq, termDocumentFreq, cb) {
-        // For each unique word in doc; add to termDocumentFreq dict or add one to its frequency
+    /*
+    Update document frequency dict; if haven't seen before, add to dict. 
+    If have, increase frequency + 1
+    Params: 
+    - Term frequency dictionary: {}
+    - Document frequency dictionary: {}
+    - cb: function()
+    */
+    updateDocumentFreq(termFreq, documentFreq, cb) {
```

Comment:
Wait, why does this accept a callback?
Code seems to be synchronous.
If so, remove cb() from param.

Caller can just
updateDocumentFreq();
cb();

Reply:
😮 👍